libs/cardano-ledger-test/src/Test/Cardano/Ledger/Generic/GenState.hs

```diff
@@ -101,66 +116,114 @@ data GenSize = GenSize
+      numPools = 5,
+      treasury = 1000000,
+      reserves = 1000000,
+      slot = 10000000
+    }
+
+getSlot :: GenState era -> SlotNo
+getSlot = SlotNo . slot . geSize . gsGenEnv
+
+getTreasury :: GenState era -> Coin
+getTreasury = Coin . treasury . geSize . gsGenEnv
+
+getReserves :: GenState era -> Coin
+getReserves = Coin . reserves . geSize . gsGenEnv
+
+getPrevUtxoElem :: GenR era (Maybe (TxIn (Crypto era), Core.TxOut era))
+getPrevUtxoElem = do
+  (UTxO x) <- Reader.asks gsPrevUTxO
+  lift $ genMapElem (SplitMap.toMap x)
+
+getPrevUtxoTest :: GenR era (TxIn (Crypto era) -> Bool)
+getPrevUtxoTest = do
+  (UTxO x) <- Reader.asks gsPrevUTxO
+  pure (\k -> maybe False (const True) (SplitMap.lookup k x))
 
 -- ========================================================================
 -- Tools to get started
 
-initState :: Reflect era => Proof era -> GenSize -> GenRS era (GenState era)
+initState :: Reflect era => Proof era -> GenSize -> GenRS era ()
 initState proof gs = do
   replicateM_ (numKeys gs) genKeyHash
   replicateM_ (numScripts gs) $ sequence_ [genScript proof i | i <- [Spend, Tag.Mint, Cert, Rewrd]]
   replicateM_ (numDatums gs) genDatumWithHash
   replicateM_ (numRewards gs) genRewards
   replicateM_ (numPools gs) genPool
-  get
-
-startGen :: Reflect era => Proof era -> GenSize -> Gen (GenEnv era, GenState era)
-startGen proof gsize = do
-  env <- genGenEnv proof
-  (state, _, _) <- runRWST (initState proof gsize) env emptyGenState
-  pure (env, state)
 
-startGenRS :: Reflect era => Proof era -> GenSize -> GenRS era (GenEnv era, GenState era)
-startGenRS proof gsize = lift $ startGen proof gsize
+genFreshState :: Reflect era => Proof era -> GenSize -> Gen (GenState era)
+genFreshState proof gsize = do
+  env <- genGenEnv proof gsize
+  (_, state, _) <- runRWST (initState proof gsize) env (emptyGenState proof env)
+  pure state
 
--- | Helper function for debugging the state generator inside ghci
-viewGenState :: Reflect era => Proof era -> IO ()
-viewGenState proof = do
-  (_, st) <- generate (startGen proof def)
-  putStrLn (show (ppGenState proof st))
+viewGenState :: Reflect era => Proof era -> Bool -> IO ()
+viewGenState proof verbose = do
+  st <- generate (genFreshState proof small)
+  when verbose $ putStrLn (show (ppGenState proof st))
+  putStrLn (show (genStateSummary proof st))
 
 -- ===========================================================
 
 -- | Generate a random, well-formed, GenEnv
-genGenEnv :: Proof era -> Gen (GenEnv era)
-genGenEnv proof = do
-  maxTxExUnits <- arbitrary :: Gen ExUnits
-  Positive maxCollateralInputs <- arbitrary :: Gen (Positive Natural)
-  collateralPercentage <- fromIntegral <$> chooseInt (1, 10000) :: Gen Natural
+genGenEnv :: Proof era -> GenSize -> Gen (GenEnv era)
+genGenEnv proof gsize = do
+  maxTxExUnits <- (arbitrary :: Gen ExUnits)
+  maxCollateralInputs <- elements [1 .. 7 :: Natural]
+  collateralPercentage <- (fromIntegral <$> chooseInt (1, 10000)) :: Gen Natural
```

Comment:
any reason to add these parens?

Reply:
When ever I add a 'typing' I do this because it is easy to get this wrong.

consider:   foo x :: Int
does it mean:   (foo x)::Int    or  (foo (x::int))

In this case it is just a variable, so it probably doesn't matter.

Reply:
It is fixed